### ISSUE TITLE
PTO handling

### ISSUE THREAD

**Issue by djc** [djc/quinn#815]
```rust
fn set_key_discard_timer(&mut self) {
    let duration = self.pto() * 3;
}

fn detect_lost_packets(&mut self) {
    let rtt = self.path.rtt.get();
    let loss_delay = cmp::max(rtt.mul_f32(self.config.time_threshold), TIMER_GRANULARITY);
    let congestion_period = self.pto() * self.config.persistent_congestion_threshold;
}

fn reset_idle_timeout(&mut self) {
     let duration = cmp::max(_, 3 * self.pto());
}

fn migrate(&self) {
    let duration = 3 * cmp::max(self.pto(), 2 * self.config.initial_rtt);
}

fn set_close_timer(&mut self) {
     let duration = 3 * self.pto();
}

fn pto_time_and_space(&self) -> Option<Duration> {
    let srtt = self.path.rtt.smoothed.unwrap_or(self.config.initial_rtt);
    let backoff = 2u32.pow(self.pto_count.min(MAX_BACKOFF_EXPONENT));
    let mut duration = (srtt + TIMER_GRANULARITY.max(4 * self.path.rtt.var)) * backoff;

    if space_id = SpaceId::Data {
        duration += self.max_ack_delay() * backoff;
    }
}

fn pto(&self) -> Duration {
    match self.path.rtt.pto_base() {
        None => 2 * self.config.initial_rtt,
        Some(base) => base + self.max_ack_delay(),
    }
}

fn pto_base(&self) -> Option<Duration> {
    self.smoothed
        .map(|srtt| srtt + cmp::max(4 * self.var, TIMER_GRANULARITY))
}

fn get(&self) -> Duration {
    self.smoothed
        .map_or(self.latest, |x| cmp::max(x, self.latest))
}
```

* Should we name the concept of `3 * PTO`?
* Is `latest` better than `initial_rtt`?
* Should we use `initial_rtt` or `2 * initial_rtt`?
* Are we systematic enough about applying `max_ack_delay`?
* Maybe `RttEstimator` should hold an `Arc<TransportConfig>` to make abstraction easier?

It would probably be good to gather all this stuff in one place to make sure we updates are consistent.

**Comment by Ralith** [comment]
Apologies for the delay reviewing this! It brought my attention to a few issues and inconsistencies which I've opened https://github.com/djc/quinn/pull/822 to address.

> Should we name the concept of 3 * PTO?

That seems like a sensible idea. Maybe `conservative_delivery_period`? Bit wordy...

> Is latest better than initial_rtt?

These are categorically different. `initial_rtt` is only used when we have no data, or as a lower bound for the path validation timeout.

> Should we use initial_rtt or 2 * initial_rtt?

For what? The spec recommends different durations for different tasks. See e.g. [the path validation spec](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#name-failed-path-validation). In general, it looks like `initial_rtt` is used as a fallback value for an RTT estimate when no data is available, and `2 * initial_rtt` is used as a lower bound for certain timer intervals which are intended to last longer than an RTT. Although on review I don't see where I got that for the PTO... maybe the spec changed.

> Are we systematic enough about applying max_ack_delay?

Where might it be missing?

> Maybe RttEstimator should hold an Arc<TransportConfig> to make abstraction easier?

This seems like overkill to me; passing in one value in `update` isn't a big deal.

**Comment by djc** [comment]
Update (after #823):

```rust
fn set_key_discard_timer() {
     let duration = self.pto() * 3;
}

fn detect_lost_packets() {
    let rtt = self.path.rtt.conservative();
    let loss_delay = cmp::max(rtt.mul_f32(self.config.time_threshold), TIMER_GRANULARITY);

    let congestion_period = self.pto() * self.config.persistent_congestion_threshold;
}

fn reset_idle_timeout() {
    let dt = cmp::max(_, 3 * self.pto());
}

fn migrate() {
    let duration = 3 * cmp::max(self.pto(), 2 * self.config.initial_rtt);
}

fn set_close_timer() {
    let duration = 3 * self.pto();
}

fn pto_time_and_space() {
    let backoff = 2u32.pow(self.pto_count.min(MAX_BACKOFF_EXPONENT));
    let mut duration = self.path.rtt.pto_base() * backoff;

    if space == SpaceId::Data {
        duration += self.max_ack_delay() * backoff;
    }
}

fn pto() {
    self.path.rtt.pto_base() + self.max_ack_delay()
}
```

So my remaining questions I think boil down to:

* Are there good reasons to bound the PTO by `2 * initial_rtt` for `migrate()`, but not `set_key_discard_timer()`, `reset_idle_timeout()` or `set_close_timer()`?
* `pto_time_and_space()` takes care to only add the `max_ack_delay()` to the PTO for `SpaceId::Data`, but `set_key_discard_timer()`, `detect_lost_packets()`, `reset_idle_timeout()`, `migrate()` and `set_close_timer()` use `pto()` directly, which always adds the `max_ack_delay()`. Presumably at least some of these can be called in earlier spaces. Should we add a `SpaceId` argument to `pto()`?

**Comment by Ralith** [comment]
> Are there good reasons to bound the PTO to 2 * initial_rtt for migrate(), but not set_key_discard_timer(), reset_idle_timeout() or set_close_timer()?

Yes. This is a lower bound, which ensures that migrating to a path with much higher latency (e.g. going from a LAN to internet connection to the same server) is possible within reasonable bounds.

> Should we add a SpaceId argument to pto()?

[Computing PTO](https://quicwg.org/base-drafts/draft-ietf-quic-recovery.html#name-computing-pto):
> `PTO = smoothed_rtt + max(4*rttvar, kGranularity) + max_ack_delay`
>
> ... When the PTO is armed for Initial or Handshake packet number spaces, the max_ack_delay is 0

This seems to support that idea. Thinking through the logic, though, I'm not sure we should *ever* be considering max_ack_delay for some of these; its purpose is to avoid acting prematurely when an ACK might still be expected, but many of those timers have nothing to do with ACKs at all: for example, path verification done by `migrate` elicits an immediate response, and the close and key discard timers are more or less indifferent.

**Comment by djc** [comment]
Going through those one by one:

* `set_key_discard_timer()`: spec on [Discarding Keys and Packet State](https://quicwg.org/base-drafts/draft-ietf-quic-recovery.html#name-discarding-keys-and-packet-) says "It is expected that keys are discarded after packets encrypted with them would be acknowledged or declared lost.", which I think supports the idea that key discarding should consider the ACK delay.

* `detect_lost_packets()`: spec on [Time Treshold](https://quicwg.org/base-drafts/draft-ietf-quic-recovery.html#time-threshold) for detecting packet loss says the time threshold is `max(kTimeThreshold * max(smoothed_rtt, latest_rtt), kGranularity)`, so it does not include the ACK delay. This seems surprising to me, as the ACK delay seems highly relevant in this case...

* `reset_idle_timeout()`: spec on [Idle Timeout](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#name-idle-timeout) says "To avoid excessively small idle timeout periods, endpoints MUST increase the idle timeout period to be at least three times the current Probe Timeout (PTO).", so I think the ACK delay should be included here.

* `migrate()`: spec on [Failed Path Validation](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#name-failed-path-validation) says `validation_timeout = max(3*PTO, 6*kInitialRtt)`, so ACK delay should be included here.

* `set_close_timer()`: spec on [Closing And Draining Connection States](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#name-closing-and-draining-connec) says "These states SHOULD persist for at least three times the current Probe Timeout (PTO) interval as defined in [QUIC-RECOVERY].", so I think this should keep the ACK delay as well.

**Comment by larseggert** [comment]
You might want to raise this as an issue over on https://github.com/quicwg/base-drafts

**Comment by djc** [comment]
Filed https://github.com/quicwg/base-drafts/issues/3951.
